### PR TITLE
Add Ceph nano for minimal object storage

### LINKED ISSUE #149: Add Object Storage component
Add a component to deploy object storage for use in ODH deployments

This is a followup item from the discussion in #104. We won't add OCS as the object storage provider as it is a downstream project. Rook will be the initial candidate and we will need to provide customizations/overlays for the default 3 storage node requirement or a smaller single node deployment

### PR DESCRIPTION
**Pull request by LaVLaS**
Closes #149 

Verification steps:
1. Use `http://ceph-nano-0` as the s3_endpoint_url
1. Retrieve the s3 credentials following the steps outlined in the readme
1. Start a jupyter notebook and use boto to verify access as outlined in the first 3 cells of the [data engineering notebook](https://gitlab.com/opendatahub/data-engineering-and-machine-learning-workshop/-/blob/master/source/notebooks/hybrid-data-engineering.ipynb

```
import sys
import os
import boto3


s3_endpoint_url = os.environ['S3_ENDPOINT_URL']
s3_access_key = os.environ['AWS_ACCESS_KEY_ID']
s3_secret_key = os.environ['AWS_SECRET_ACCESS_KEY']
s3_bucket_name = os.environ['JUPYTERHUB_USER']

print(s3_endpoint_url)
print(s3_bucket_name)
s3 = boto3.client('s3','us-east-1', endpoint_url= s3_endpoint_url,
                       aws_access_key_id = s3_access_key,
                       aws_secret_access_key = s3_secret_key)
  Creating a bucket, uploading an object (put), and listing the bucket.
 s3.create_bucket(Bucket=s3_bucket_name)
s3.put_object(Bucket=s3_bucket_name,Key='object',Body='data')
for key in s3.list_objects(Bucket=s3_bucket_name)['Contents']:
    print(key['Key'])
```

### CODE REVIEW

File: ceph/object-storage/nano/base/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+- name: namespace
+  objref:
+    kind: ConfigMap
+    name: ceph-nano-parameters
+    apiVersion: v1
+  fieldref:
+    fieldpath: metadata.namespace
+
+configurations:
+- params.yaml
```
Comment (vpavlin) [inline ceph/object-storage/nano/base/kustomization.yaml:33]: missing new line

File: ceph/object-storage/nano/base/persistentvolumeclaim.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    requests:
+        storage: 10Gi
+  # Uncomment and add storageClass specific to your requirements below. Read more https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1
+  #storageClassName:
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: cn-pv-claim-etc
+  namespace: $(namespace)
```
Comment (vpavlin) [inline ceph/object-storage/nano/base/persistentvolumeclaim.yaml:23]: I believe namespace is added by default - you don't need to explicitely add the variable

**Comment by vpavlin** [comment]
@LaVLaS I don't think you need to use the `$(namespace)` variable in `metadata.namespace` - that is added/replaced by Kustomize by default. Only the non-standard locations need to be templated

File: ceph/object-storage/nano/base/persistentvolumeclaim.yaml (hunk 2 above)
Comment (LaVLaS) [reply ceph/object-storage/nano/base/persistentvolumeclaim.yaml]: Updated.

File: ceph/object-storage/nano/base/kustomization.yaml (hunk 1 above)
Comment (LaVLaS) [reply ceph/object-storage/nano/base/kustomization.yaml:39]: Missing new line where?

Comment (nakfour) [reply ceph/object-storage/nano/base/kustomization.yaml:39]: after -params.yaml

**Comment by nakfour** [comment]
This worked for me. @LaVLaS just your example code has a comment with no "#"

File: ceph/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Ceph Object Storage
+
+[Ceph](https://ceph.io) provides S3 compatible object storage via the RADOS GateWay(RGW) Service
+
+Deploys [cn-core](https://github.com/cn-core) image that provides minimal object storage for small deployments based on a release of [ceph-container](https://github.com/ceph/ceph-container).
+
+***NOTE***: This container MUST run as a PRIVILEGED container with anyuid enabled
+
```
Comment (nakfour) [inline ceph/README.md:8]: Is there a way around this?

File: ceph/object-storage/nano/base/clusterrolebinding.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    app: ceph
+  name: cluster-admin-ceph-nano
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cluster-admins
+subjects:
+- kind: ServiceAccount
+  name: ceph-nano
+  namespace: $(namespace)
```
Comment (nakfour) [inline ceph/object-storage/nano/base/clusterrolebinding.yaml:14]: Why does it need a clusterrole?

File: ceph/object-storage/nano/base/service.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    daemon: nano
+spec:
+  ports:
+  - name: cn-sree
+    port: 5001
+    protocol: TCP
+    targetPort: 5001
+  selector:
+    app: ceph
+    daemon: nano
```
Comment (nakfour) [inline ceph/object-storage/nano/base/service.yaml:52]: Is there a prometheus metrics port?

**Comment by nakfour** [comment]
Also ideally we add tests for new components, you can just open and issue task to do that later @LaVLaS

**Comment by nakfour** [comment]
Ceph-nano portal has issues, uploading to a bucket caused an error.

**Comment by crobby** [comment]
I've added the ceph chunk to my kfdef and pointed the manifests to this PR, but when I fire-up the KfDef, I see "Encountered error applying application ceph-nano:  (kubeflow.error): Code 500 with message: Apply.Run : no objects passed to apply" in my operator log.

Here is my snippet from the KfDef:
```
  - kustomizeConfig:
      repoRef:
        name: manifests
        path: ceph/object-storage/nano
    name: ceph-nano
```

File: ceph/README.md (hunk 3 above)
Comment (LaVLaS) [reply ceph/README.md:8]: I removed the Privilege execution. It only requires anyuid permissions now

File: ceph/object-storage/nano/base/clusterrolebinding.yaml (hunk 4 above)
Comment (LaVLaS) [reply ceph/object-storage/nano/base/clusterrolebinding.yaml]: It doesn't the rolebinding was removed

File: ceph/object-storage/nano/base/service.yaml (hunk 5 above)
Comment (LaVLaS) [reply ceph/object-storage/nano/base/service.yaml]: It doesn't appear that metrics are enabled by default (if available)

File: ceph/object-storage/nano/base/kustomization.yaml (hunk 1 above)
Comment (LaVLaS) [reply ceph/object-storage/nano/base/kustomization.yaml]: Updated.

We should add a CI job that will enforce this new style requirement for all yaml files

**Comment by crobby** [comment]
This is working for me.  I approve.  I'll let someone else do the magic lgtm.

**Pull request merged**